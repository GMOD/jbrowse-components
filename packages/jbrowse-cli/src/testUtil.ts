/* eslint-disable import/no-extraneous-dependencies */
import { test as oclifTest } from '@oclif/test'
import del from 'del'
import fs from 'fs'
import os from 'os'
import path from 'path'

const fsPromises = fs.promises

// On macOS, os.tmpdir() is not a real path: https://github.com/nodejs/node/issues/11422
const tmpDir = fs.realpathSync(os.tmpdir())

/* eslint-disable no-console */
const mockConsoleLog = (opts?: { print?: boolean }) => ({
  run(ctx: { consoleLog: jest.Mock }) {
    const originalConsoleLog = console.log.bind(console)
    const consoleLogMock = jest.fn((...data: unknown[]) => {
      if (opts?.print) {
        originalConsoleLog(...data)
      }
    })
    console.log = consoleLogMock
    ctx.consoleLog = console.log as typeof consoleLogMock
  },
})
/* eslint-enable no-console */

const mockConsoleWarn = (opts?: { print?: boolean }) => ({
  run(ctx: { consoleWarn: jest.Mock }) {
    const originalConsoleWarn = console.warn.bind(console)
    const consoleWarnMock = jest.fn((...data: unknown[]) => {
      if (opts?.print) {
        originalConsoleWarn(...data)
      }
    })
    console.warn = consoleWarnMock
    ctx.consoleWarn = console.warn as typeof consoleWarnMock
  },
})

const mockConsoleError = (opts?: { print?: boolean }) => ({
  run(ctx: { consoleError: jest.Mock }) {
    const originalConsoleError = console.error.bind(console)
    const consoleErrorMock = jest.fn((...data: unknown[]) => {
      if (opts?.print) {
        originalConsoleError(...data)
      }
    })
    console.error = consoleErrorMock
    ctx.consoleError = console.error as typeof consoleErrorMock
  },
})

const mockStdoutWrite = (opts?: { print?: boolean }) => ({
  run(ctx: { stdoutWrite: jest.Mock }) {
    const stdoutWriteMock = jest.fn((data: Buffer | string) => {
      if (opts?.print) {
        // eslint-disable-next-line no-console
        console.log(data)
      }
      return true
    })
    process.stdout.write = stdoutWriteMock
    ctx.stdoutWrite = process.stdout.write as typeof stdoutWriteMock
  },
})

export const test = oclifTest
  .register('mockConsoleLog', mockConsoleLog)
  .register('mockConsoleWarn', mockConsoleWarn)
  .register('mockConsoleError', mockConsoleError)
  .register('mockStdoutWrite', mockStdoutWrite)

export const setup = test
  .mockStdoutWrite()
  .add('originalDir', () => process.cwd())
  .add('dir', async () => {
    const jbrowseTmpDir = path.join(tmpDir, 'jbrowse')
    await fsPromises.mkdir(jbrowseTmpDir, { recursive: true })
    return fsPromises.mkdtemp(path.join(jbrowseTmpDir, path.sep))
  })
  .finally(async ctx => {
    await del([`${ctx.dir}/**`, ctx.dir], { force: true })
    process.chdir(ctx.originalDir)
  })
  .do(async ctx => {
    process.chdir(ctx.dir)
    await fsPromises.writeFile('manifest.json', '{"name":"JBrowse"}')
  })
