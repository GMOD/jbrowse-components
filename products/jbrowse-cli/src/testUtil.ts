import fs from 'fs'
import os from 'os'
import path from 'path'
import { runCommand } from '@oclif/test'

const { mkdir, mkdtemp } = fs.promises

// increase test timeout for all tests
jest.setTimeout(20000)

// On macOS, os.tmpdir() is not a real path:
// https://github.com/nodejs/node/issues/11422
const tmpDir = fs.realpathSync(os.tmpdir())

export async function runInTmpDir(
  callbackFn: (args: { dir: string; originalDir: string }) => Promise<void>,
) {
  const originalDir = process.cwd()
  let dir: string | undefined
  try {
    const jbrowseTmpDir = path.join(tmpDir, 'jbrowse')
    await mkdir(jbrowseTmpDir, { recursive: true })
    dir = await mkdtemp(path.join(jbrowseTmpDir, path.sep))
    process.chdir(dir)
    await callbackFn({ dir, originalDir })
  } finally {
    if (dir) {
      fs.rmSync(dir, { recursive: true, force: true })
    }
    process.chdir(originalDir)
  }
}
export async function setup(str: string | string[]) {
  return runCommand(str)
}

type Conf = Record<string, any>

export function readConf(ctx: { dir: string }, ...rest: string[]): Conf {
  return JSON.parse(
    fs.readFileSync(path.join(ctx.dir, ...rest, 'config.json'), 'utf8'),
  )
}

export function readConfAlt(ctx: { dir: string }, ...rest: string[]): Conf {
  return JSON.parse(fs.readFileSync(path.join(ctx.dir, ...rest), 'utf8'))
}

export function dataDir(str: string) {
  return path.join(__dirname, '..', 'test', 'data', str)
}

export function ctxDir(ctx: { dir: string }, str: string) {
  return path.join(ctx.dir, str)
}
