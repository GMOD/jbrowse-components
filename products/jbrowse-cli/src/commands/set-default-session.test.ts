/**
 * @jest-environment node
 */

import fs from 'fs'
import * as path from 'path'
import { setup } from '../testUtil'

const fsPromises = fs.promises

const defaultConfig = {
  assemblies: [
    {
      name: 'testAssembly',
      sequence: {
        type: 'testSequenceTrack',
        trackId: '',
        adapter: {
          type: 'testSeqAdapter',
          twoBitLocation: {
            uri: 'test.2bit',
          },
        },
      },
    },
  ],
  configuration: {},
  connections: [],
  defaultSession: {
    name: 'New Session',
  },
  tracks: [
    {
      type: 'AlignmentsTrack',
      trackId: 'simple',
      name: 'simple',
      assemblyNames: ['testAssembly'],
      adapter: {
        type: 'BamAdapter',
        bamLocation: {
          uri: 'simple.bam',
        },
        index: {
          location: {
            uri: 'simple.bam.bai',
          },
        },
        sequenceAdapter: {
          type: 'testSeqAdapter',
          twoBitLocation: {
            uri: 'test.2bit',
          },
        },
      },
    },
  ],
}

const simpleBam = path.join(__dirname, '..', '..', 'test', 'data', 'simple.bam')
const simpleDefaultSession = path.join(
  __dirname,
  '..',
  '..',
  'test',
  'data',
  'sampleDefaultSession.json',
)

const testConfig = path.join(
  __dirname,
  '..',
  '..',
  'test',
  'data',
  'test_config.json',
)

const setupWithAddTrack = setup
  .do(async ctx => {
    await fsPromises.copyFile(
      testConfig,
      path.join(ctx.dir, path.basename(testConfig)),
    )

    await fsPromises.rename(
      path.join(ctx.dir, path.basename(testConfig)),
      path.join(ctx.dir, 'config.json'),
    )
  })
  .command(['add-track', simpleBam, '--load', 'copy'])

describe('set-default-session', () => {
  setup
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command(['set-default-session'])
    .exit(120)
    .it('fails when no necessary default session information is provided')
  setup
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command(['set-default-session', '--session', '{}'])
    .exit(150)
    .it('fails when default session is not readable')
  setup
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command([
      'set-default-session',
      '--session',
      path.join(simpleDefaultSession, 'nonexist.json'),
    ])
    .exit(150)
    .it('fails when file does not exist')
  setup
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command(['set-default-session', '--session', simpleBam])
    .exit(160)
    .it('fails when file is does not have a default session to read')
  setup
    .do(async () => {
      await fsPromises.unlink('manifest.json')
    })
    .command(['set-default-session', '--session', simpleDefaultSession])
    .exit(10)
    .it('fails if no manifest.json found in cwd')
  setup
    .do(async () => {
      await fsPromises.writeFile('manifest.json', 'This Is Invalid JSON')
    })
    .command(['set-default-session', '--session', simpleDefaultSession])
    .exit(20)
    .it("fails if it can't parse manifest.json")

  setup
    .do(async () => {
      await fsPromises.writeFile('manifest.json', '{"name":"NotJBrowse"}')
    })
    .command(['set-default-session', '--session', simpleDefaultSession])
    .exit(30)
    .it('fails if "name" in manifest.json is not "JBrowse"')
  setupWithAddTrack
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command(['set-default-session', '--tracks', 'simple'])
    .exit(130)
    .it('fails when specifying a track without specifying a view')
  setupWithAddTrack
    .command([
      'set-default-session',
      '--view',
      'LinearGenomeView',
      '--tracks',
      'track-non-exist',
    ])
    .exit(140)
    .it('fails when specifying a track that does not exist')
  setup
    .do(async ctx => {
      await fsPromises.copyFile(
        testConfig,
        path.join(ctx.dir, path.basename(testConfig)),
      )

      await fsPromises.rename(
        path.join(ctx.dir, path.basename(testConfig)),
        path.join(ctx.dir, 'config.json'),
      )
    })
    .command(['set-default-session', '--session', simpleDefaultSession])
    .it('adds a default session from a file', async ctx => {
      const contents = await fsPromises.readFile(
        path.join(ctx.dir, 'config.json'),
        { encoding: 'utf8' },
      )
      expect(JSON.parse(contents)).toEqual({
        ...defaultConfig,
        tracks: [],
        defaultSession: {
          name: 'test new session',
          views: [
            {
              id: '823WX',
              type: 'LinearGenomeView',
              tracks: [
                {
                  type: 'AlignmentsTrack',
                  configuration: 'simple',
                },
              ],
            },
          ],
        },
      })
    })
  setupWithAddTrack
    .command([
      'set-default-session',
      '--view',
      'LinearGenomeView',
      '--tracks',
      'simple',
    ])
    .it(
      'adds a default session that is a linear genome view and a simple track',
      async ctx => {
        const contents = await fsPromises.readFile(
          path.join(ctx.dir, 'config.json'),
          { encoding: 'utf8' },
        )
        expect(JSON.parse(contents)).toEqual({
          ...defaultConfig,
          defaultSession: {
            name: 'New Default Session',
            views: [
              {
                id: 'LinearGenomeView-1',
                type: 'LinearGenomeView',
                tracks: [
                  {
                    type: 'AlignmentsTrack',
                    configuration: 'simple',
                  },
                ],
              },
            ],
          },
        })
      },
    )
})
