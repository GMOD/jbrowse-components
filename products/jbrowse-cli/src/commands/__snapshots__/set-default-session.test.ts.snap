// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a default session from a file 1`] = `
{
  "assemblies": [
    {
      "name": "testAssembly",
      "sequence": {
        "adapter": {
          "twoBitLocation": {
            "locationType": "UriLocation",
            "uri": "test.2bit",
          },
          "type": "testSeqAdapter",
        },
        "trackId": "",
        "type": "testSequenceTrack",
      },
    },
  ],
  "configuration": {},
  "connections": [],
  "defaultSession": {
    "name": "test new session",
    "views": [
      {
        "id": "823WX",
        "tracks": [
          {
            "configuration": "simple",
            "type": "AlignmentsTrack",
          },
        ],
        "type": "LinearGenomeView",
      },
    ],
  },
  "tracks": [],
}
`;

exports[`deletes a default session 1`] = `
{
  "assemblies": [
    {
      "name": "testAssembly",
      "sequence": {
        "adapter": {
          "twoBitLocation": {
            "locationType": "UriLocation",
            "uri": "test.2bit",
          },
          "type": "testSeqAdapter",
        },
        "trackId": "",
        "type": "testSequenceTrack",
      },
    },
  ],
  "configuration": {},
  "connections": [],
  "tracks": [],
}
`;

exports[`fails when default session is not readable 1`] = `"Could not read the provided file"`;

exports[`fails when file does not exist 1`] = `"Could not read the provided file"`;

exports[`fails when file is does not have a default session to read 1`] = `"Could not parse the given default session file"`;

exports[`fails when no necessary default session information is provided 1`] = `"Please provide a --session file"`;
